### PR TITLE
Sync execution update on demand

### PR DESCRIPTION
**Pull request by yrong**
Requires: https://github.com/Snowfork/polkadot-sdk/pull/123

### CODE REVIEW

File: relayer/relays/beacon/header/header.go [hunk 1]
```diff
@@ -208,79 +204,12 @@ func (h *Header) SyncHeaders(ctx context.Context) error {
...
-}
-
 func (h *Header) syncLaggingSyncCommitteePeriods(ctx context.Context, latestSyncedPeriod, currentSyncPeriod uint64) error {
-	// sync for the next period
-	periodsToSync := []uint64{latestSyncedPeriod + 1}
+	// sync for all missing periods
+	periodsToSync := []uint64{}
+	for i := latestSyncedPeriod + 1; i <= currentSyncPeriod; i++ {
+		periodsToSync = append(periodsToSync, i)
+	}
```
[the remark this answers is not in the document]
Comment (yrong) [reply relayer/relays/beacon/header/header.go:212]: > The sync committee is only used to check finalized beacon headers and not execution headers.

Exactly! I think it should most probably work with https://github.com/Snowfork/snowbridge/pull/1153 though I closed.

[the remark this answers is not in the document]
Comment (claravanstaden) [reply relayer/relays/beacon/header/header.go:212]: I think so, it's a useful change. 😄

File: relayer/relays/beacon/header/syncer/scale/beacon_scale.go [hunk 2]
```diff
@@ -41,6 +41,26 @@ type UpdatePayload struct {
 	FinalityBranch          []types.H256
 	BlockRootsRoot          types.H256
 	BlockRootsBranch        []types.H256
+	ExecutionHeader         OptionExecutionHeader
+	ExecutionBranch         OptionExecutionBranch
```
[the remark this answers is not in the document]
Comment (yrong) [reply relayer/relays/beacon/header/syncer/scale/beacon_scale.go]: https://github.com/Snowfork/snowbridge/pull/1154/commits/ddf7e6a43f19275906a28e06ef515ec71fd6f5c5

File: relayer/relays/execution/config.go [hunk 3]
```diff
@@ -2,6 +2,7 @@ package execution
 
 import (
 	"github.com/snowfork/snowbridge/relayer/config"
+	BeaconConfigs "github.com/snowfork/snowbridge/relayer/relays/beacon/config"
```
[the remark this answers is not in the document]
Comment (yrong) [reply relayer/relays/execution/config.go]: https://github.com/Snowfork/snowbridge/pull/1154/commits/cd85cb9a1a5aad852c520213f18e103eee4a7dc2

File: relayer/relays/beacon/header/header.go [hunk 4]
```diff
@@ -353,6 +282,13 @@ func (h *Header) populateClosestCheckpoint(slot uint64) (cache.Proof, error) {
 		checkpointSlot := checkpoint.Slot
 		if checkpointSlot == 0 {
 			checkpointSlot = h.syncer.CalculateNextCheckpointSlot(slot)
+			lastFinalizedHeaderState, err := h.writer.GetLastFinalizedHeaderState()
+			if err != nil {
+				return cache.Proof{}, fmt.Errorf("fetch parachain last finalized header state: %w", err)
+			}
+			if checkpointSlot > lastFinalizedHeaderState.BeaconSlot {
+				checkpointSlot = lastFinalizedHeaderState.BeaconSlot
+			}
```
[the remark this answers is not in the document]
Comment (yrong) [reply relayer/relays/beacon/header/header.go]: https://github.com/Snowfork/snowbridge/pull/1154/commits/03bdb5c6e31ac3b0f0b46e414a58a38ca93aad33

File: relayer/relays/beacon/header/header.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (yrong) [reply relayer/relays/beacon/header/header.go:212]: https://github.com/Snowfork/snowbridge/pull/1154/commits/ef8c27aa5ce4bd21640d239dc2e8a740beae1116

File: relayer/relays/execution/main.go [hunk 5]
```diff
@@ -145,6 +146,22 @@ func (r *Relay) Start(ctx context.Context, eg *errgroup.Group) error {
...
+				if err != nil {
+					return fmt.Errorf("get block header: %w", err)
+				}
+
+				// ParentBeaconRoot in https://eips.ethereum.org/EIPS/eip-4788 from Deneb onward
+				err = beaconHeader.SyncExecutionHeader(ctx, *blockHeader.ParentBeaconRoot)
+				if err == header.ErrBeaconHeaderNotFinalized {
+					logger.Warn("beacon header not finalized, just skipped")
+					continue
+				}
```
Comment (yrong) [inline relayer/relays/execution/main.go:161]: Just continue and wait here until that execution header of the event log is finalized.

**Review by vgeddes — APPROVED** [review]
+1

File: relayer/chain/parachain/writer.go [hunk 6]
```diff
@@ -385,3 +360,91 @@ func (wr *ParachainWriter) getNumberFromParachain(pallet, storage string) (uint6
...
+			break
+		}
+		if beaconState.BeaconSlot > slot && beaconState.BeaconSlot < slot+8192 {
+			break
+		}
+	}
+	if beaconState.BeaconSlot > slot && beaconState.BeaconSlot < slot+8192 {
+		return beaconState, nil
+	}
+	return beaconState, fmt.Errorf("Can't find checkpoint on chain for slot %d", slot)
```
Comment (claravanstaden) [inline relayer/chain/parachain/writer.go:449]:
```suggestion
	return beaconState, fmt.Errorf("find checkpoint on chain for slot %d", slot)
```

The error message is appended continuously, like here https://github.com/Snowfork/snowbridge/pull/1154/files#diff-92cc554d5b7218e7576839ca25a3b71ae0bbfa8200552beaf703a95a1a0da2a3R292, so simplify and lowercase the error message.

File: relayer/chain/parachain/writer.go [hunk 7]
```diff
@@ -385,3 +360,91 @@ func (wr *ParachainWriter) getNumberFromParachain(pallet, storage string) (uint6
...
+
+	_, err = wr.conn.API().RPC.State.GetStorageLatest(key, &beaconRoot)
+	if err != nil {
+		return beaconRoot, fmt.Errorf("get storage for FinalizedBeaconStateMapping (err): %w", err)
+	}
+
+	return beaconRoot, nil
+}
+
+func (wr *ParachainWriter) FindCheckPointBackward(slot uint64) (state.FinalizedHeader, error) {
```
Comment (claravanstaden) [inline relayer/chain/parachain/writer.go:419]:
```suggestion
func (wr *ParachainWriter) FindCheckPoint(slot uint64) (state.FinalizedHeader, error) {
```

File: relayer/chain/parachain/writer.go [hunk 8]
```diff
@@ -385,3 +360,91 @@ func (wr *ParachainWriter) getNumberFromParachain(pallet, storage string) (uint6
...
+			return beaconState, fmt.Errorf("GetFinalizedBeaconRootByIndex %d, error: %w", index, err)
+		}
+		beaconState, err = wr.GetFinalizedHeaderStateByBlockRoot(beaconRoot)
+		if err != nil {
+			return beaconState, fmt.Errorf("GetFinalizedHeaderStateByBlockRoot %s, error: %w", beaconRoot.Hex(), err)
+		}
+		if beaconState.BeaconSlot < slot {
+			break
+		}
+		if beaconState.BeaconSlot > slot && beaconState.BeaconSlot < slot+8192 {
```
Comment (claravanstaden) [inline relayer/chain/parachain/writer.go:442]: Should rather use `SLOTS_PER_EPOCH * EPOCHS_PER_SYNC_COMMITTEE_PERIOD` here. Please also add code comments explaining this line and the one above, it's not clear to me what these 2 checks check.

File: relayer/relays/beacon/header/header.go [hunk 9]
```diff
@@ -346,39 +269,57 @@ func (h *Header) populateFinalizedCheckpoint(slot uint64) error {
...
+			checkpointSlot = h.syncer.CalculateNextCheckpointSlot(slot)
+			lastFinalizedHeaderState, err := h.writer.GetLastFinalizedHeaderState()
+			if err != nil {
+				return checkpoint, fmt.Errorf("get last finalized header for the checkpoint: %w", err)
+			}
+			if slot > lastFinalizedHeaderState.BeaconSlot {
+				return checkpoint, ErrBeaconHeaderNotFinalized
+			}
+			if checkpointSlot < lastFinalizedHeaderState.BeaconSlot {
+				log.WithFields(log.Fields{"calculatedCheckpointSlot": checkpointSlot, "lastFinalizedSlot": lastFinalizedHeaderState.BeaconSlot}).Info("fetch checkpoint on chain backward from history")
```
Comment (claravanstaden) [inline relayer/relays/beacon/header/header.go:289]:
```suggestion
				log.WithFields(log.Fields{"calculatedCheckpointSlot": checkpointSlot, "lastFinalizedSlot": lastFinalizedHeaderState.BeaconSlot}).Info("fetch closest checkpoint on chain")
```

File: relayer/relays/beacon/header/header.go [hunk 10]
```diff
@@ -394,14 +335,48 @@ func (h *Header) getNextHeaderUpdateBySlot(slot uint64) (scale.HeaderUpdatePaylo
...
+	if err != nil {
+		return fmt.Errorf("get compactExecutionHeaderState by blockHash: %w", err)
+	}
+	if compactExecutionHeaderState.BlockNumber != 0 {
+		log.WithFields(log.Fields{"blockRoot": blockRoot.Hex(), "blockHash": blockHash.Hex(), "blockNumber": compactExecutionHeaderState.BlockNumber}).Info("ExecutionHeaderState already exist")
+		return nil
+	}
+	err = h.writer.WriteToParachainAndWatch(ctx, "EthereumBeaconClient.submit_execution_header", headerUpdate)
+	if err != nil {
+		return fmt.Errorf("submit_execution_header: %w", err)
```
Comment (claravanstaden) [inline relayer/relays/beacon/header/header.go:372]:
```suggestion
		return fmt.Errorf("submit execution header : %w", err)
```

**Review by claravanstaden — COMMENTED** [review]
Almost there! Looks great 😄

File: relayer/chain/parachain/writer.go (hunk 8 above)
Comment (yrong) [reply relayer/chain/parachain/writer.go:442]: It's to backward to find check point(slot<beaconState.BeaconSlot<slot+8192)

For the first check `beaconState.BeaconSlot < slot` which means there is no valid checkpoint on chain so just exit the loop.

For the second check `if beaconState.BeaconSlot > slot && beaconState.BeaconSlot < slot+8192`  which means we find that so exit the loop.

**Pull request merged**